src/dra/8D3E8.c

```diff
@@ -951,4 +951,1098 @@ s32 func_8012F83C(s32 x0, s32 y0, s32 x1, s32 y1, s32 distance) {
+#endif
+        self->flags = FLAG_UNK_04000000 | FLAG_HAS_PRIMS | FLAG_UNK_40000 |
+                      FLAG_UNK_20000;
+        self->step++;
+    }
+    animControl = 0;
+#ifndef VERSION_HD
+    self->zPriority = PLAYER.zPriority - 2;
+#endif
+    self->facingLeft = PLAYER.facingLeft;
+    if (abs(PLAYER.velocityX) > FIX(3)) {
+        PLAYER.drawFlags = self->drawFlags |= DRAW_HIDE;
+        PLAYER.drawMode = self->drawMode =
+            DRAW_UNK_40 | DRAW_TPAGE | DRAW_TPAGE2;
+
+        var_s1 = (abs(PLAYER.velocityX) - FIX(3)) >> 11;
+        if (var_s1 > 0x80) {
+            var_s1 = 0x80;
+        }
+        PLAYER.unk6C = self->unk6C = 0xFF - var_s1;
+    } else {
+        self->drawFlags &= ~DRAW_HIDE;
+        PLAYER.drawFlags = self->drawFlags;
+        PLAYER.drawMode = self->drawMode = DRAW_DEFAULT;
+    }
+    switch (PLAYER.step_s) {
+    case 1:
+        if (D_800B0914 == 1) {
+            D_800B0920 = 8;
+        }
+        break;
+    case 2:
+        switch (D_800B0914) {
+        case 0:
+            break;
+        case 1:
+        case 3:
+            D_800B0920 -= D_800B0A3C[PLAYER.animFrameIdx];
+            break;
+        case 2:
+            D_800B0920 -= D_800B0A4C[PLAYER.animFrameIdx];
+            break;
+        case 4:
+            break;
+        }
+        break;
+    case 4:
+        if (PLAYER.velocityY < 0) {
+            if (D_80138430 > 0x700) {
+                D_80138430 -= 0x20;
+            }
+            D_800B0920 = 10;
+        } else {
+            if (D_80138430 < 0x7C0) {
+                D_80138430 += 0x20;
+            }
+            D_800B0920 = 12;
+        }
+        break;
+    case 5:
+    case 8:
+        D_800B0920 = 0xC;
+        break;
+    case 7:
+        if (D_80138430 < 0x900) {
+            D_80138430 += 0x20;
+        }
+        break;
+    case 9:
+        D_800B0920 -= D_800B0A3C[PLAYER.animFrameIdx];
+        break;
+    }
+
+    playerX = PLAYER.posX.i.hi;
+    playerY = PLAYER.posY.i.hi + 25;
+    if (!(g_Player.pl_vram_flag & 1) || (PLAYER.step_s == 7)) {
+        angle = D_80138430;
+        if (PLAYER.facingLeft) {
+            angle = 0x800 - D_80138430;
+        }
+        if (g_Player.unk04 & 0x40) {
+            angle = 0x800;
+            if (PLAYER.facingLeft) {
+                angle = 0x800 - angle;
+            }
+        }
+        hitboxX = (((rcos(angle) >> 4) * D_800B0920) >> 8) + playerX;
+        hitboxY = playerY + -(((rsin(angle) >> 4) * D_800B0920) >> 8);
+    } else {
+        var_s1 = 1;
+        hitboxY = playerY + 2;
+        // This loop is really gross, I will appreciate you forever if you can
+        // make it look more reasonable.
+        while (1) {
+            if (PLAYER.facingLeft) {
+                hitboxX = playerX + var_s1;
+            } else {
+                hitboxX = playerX - var_s1;
+            }
+            hitboxY += 2;
+            CheckCollision(hitboxX, hitboxY, &collider1, 0);
+            if (collider1.effects & 1) {
```

Comment:
enum, same for other effects checks https://github.com/Xeeynamo/sotn-decomp/blob/d73363b83b55b148a350d8c979f4806447a50443/include/game.h#L1168